File: scala/src/main/scala/shapes/base/models/Rectangle.scala
package shapes.base.models

import shapes.base.commons.Dimensional

import scala.compiletime.uninitialized

final class Rectangle extends Dimensional {
    private var width: Double = uninitialized
    private var height: Double = uninitialized

    def this(width: Double, height: Double) = {
        this()
        var dimOne: Double = if (width > 0) width else 1.0
        var dimTwo: Double = if (height > 0) height else 1.0
        this.width = dimOne
        this.height = dimTwo
    }

    override def area: Double = {
        width * height
    }

    override def perimeter: Double = {
        2 * (width + height)
    }
}

File: elixir/lib/shapes/base/models/rectangle.ex
alias Shapes.Base.Commons.Dimensional
alias Shapes.Base.Models.Rectangle

defmodule Rectangle do
  defstruct width: 0.0, height: 0.0

  def new(width, height) do
    dim_one = if(width > 0, do: width, else: 1.0)
    dim_two = if(height > 0, do: height, else: 1.0)
    %__MODULE__{width: dim_one, height: dim_two}
  end

  def area(self) do
    Dimensional.area(self)
  end

  def perimeter(self) do
    Dimensional.perimeter(self)
  end
end

defimpl Dimensional, for: Rectangle do
  def area(%Rectangle{width: width, height: height}) do
    width * height
  end

  def perimeter(%Rectangle{width: width, height: height}) do
    2 * (width + height)
  end
end
